crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3797,6 +3856,33 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn narrow_singleton_pattern() {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "/src/a.py",
+            "
+            x = None if flag else 1
+            y = 0
+            match x:
+                case None:
+                    y = x
+            ",
+        )
+        .unwrap();
+
+        // TODO: The correct inferred type should be `Literal[0] | None` but currently the
+        // simplification logic doesn't account for this. The final type with parenthesis:
+        // `Literal[0] | (None | Literal[1] & None)`
```

Comment:
Thinking through the simplification we do, I'm pretty sure this is actually what we we end up with, because we normalize to disjunctive normal form (union of intersections), so `(None | 1) & None` distributes to `(None & None) | (1 & None)` which simplifies to `(None | (1 & None))`, which flattens into the outer union, giving:

```suggestion
        // `Literal[0] | None | (Literal[1] & None)`
```

The simplification we are missing here is knowing that `1 & None` is a subtype of `None`, and therefore simplifies out of the union entirely, leaving just `0 | None` in the union, which is the right answer.

(I added the wrapping parens in the union and intersection displays locally and verified what I've said here is actually correct this time!)